### ISSUE TITLE
pmap performance

### ISSUE THREAD

**Issue by rdyro** [rdyro/torch2jax#25]
Responding to: @vladyorsh from #12 

> Hi @rdyro,
> 
> can you provide instructions how to properly handle a pmap-ped traced function? I might be missing something, since in my scenario I have a stateless Torch function using a custom CUDA kernel, and I end up with the same throughput despite pmap-ping across multiple devices. Swapping the function to the pure JAX version works correctly.
> 
> Wrapper code for the function, which is a Cauchy MVM computation kernel [from there](https://github.com/state-spaces/s4/tree/main/extensions/kernels):
> ```
> import torch
> from torch2jax import torch2jax
> 
> cauchy_extension = False
> 
> try:
>   #from s4.extensions.kernels.cauchy import cauchy_mult_torch
>   #from s4.extensions.kernels.cauchy import cauchy_mult as cauchy_mult_ref
>   from structured_kernels import cauchy_mult_sym_fwd, cauchy_mult_sym_bwd
> 
>   def get_compiled_cauchy(H, L_trunc, N):
>     print(f'Providing a CUDA Cauchy mult kernel for shapes: H, L_trunc, N: {H}, {L_trunc}, {N}')
>     def cauchy_mult_ref(v, z, w):
>       v, w = torch.broadcast_tensors(v, w)
>       shape = v.shape
>       # z_shape = z.shape
>       # z = z.squeeze()
>       assert len(z.shape) == 1
> 
>       v = v.contiguous()
>       w = w.contiguous()
>       z = z.contiguous()
> 
>       N = v.size(-1)
>       assert w.size(-1) == N
>       y = cauchy_mult_sym_fwd(v.view(-1, N), z, w.view(-1, N))
>       y = y.view(*shape[:-1], z.size(-1))
>       return y.contiguous()
>       
>     def cauchy_sym(v, z, w):
>       v, z, w = torch.view_as_complex(v), torch.view_as_complex(z), torch.view_as_complex(w)
>       return torch.view_as_real(cauchy_mult_ref(v, z, w))
> 
>     cauchy_jit = jax.custom_vjp(torch2jax(
>       cauchy_sym,
>       jax.ShapeDtypeStruct((H, N, 2), jnp.float32),
>       jax.ShapeDtypeStruct((L_trunc, 2), jnp.float32),
>       jax.ShapeDtypeStruct((H, N, 2), jnp.float32),
> 
>       output_shapes=jax.ShapeDtypeStruct((H, L_trunc, 2), jnp.float32),
>     ))
> 
>     def cauchy_fw(v, z, w):
>       return cauchy_jit(v, z, w), (v, z, w)
> 
>     def cauchy_bw(v, z, w, g):
>       v, z, w, g = torch.view_as_complex(v), torch.view_as_complex(z), torch.view_as_complex(w), torch.view_as_complex(g)
>       dv, dw = cauchy_mult_sym_bwd(v, z, w, g)
>       return torch.view_as_real(dv).contiguous(), torch.view_as_real(torch.zeros_like(z)).contiguous(), torch.view_as_real(dw).contiguous()
> 
>     cauchy_bw = torch2jax(
>       cauchy_bw,
>       jax.ShapeDtypeStruct((H, N, 2), jnp.float32),
>       jax.ShapeDtypeStruct((L_trunc, 2), jnp.float32),
>       jax.ShapeDtypeStruct((H, N, 2), jnp.float32),
>       jax.ShapeDtypeStruct((H, L_trunc, 2), jnp.float32),
> 
> 
>       output_shapes=(jax.ShapeDtypeStruct((H, N, 2), jnp.float32),
>       jax.ShapeDtypeStruct((L_trunc, 2), jnp.float32),
>       jax.ShapeDtypeStruct((H, N, 2), jnp.float32),)
>     )
> 
>     @jax.jit
>     def cauchy_jit_bw(res, g):
>       v, z, w = res
>       return cauchy_bw(v, z, w, g)
> 
>     cauchy_jit.defvjp(cauchy_fw, cauchy_jit_bw)
> 
>     def r2c_wrapper(v, z, w):
>       v, z, w = c2r(v), c2r(z), c2r(w)
>       return r2c(cauchy_jit(v, z, w))
> 
>     return r2c_wrapper
> 
>   cauchy_extension = True
>   print('Loaded Cauchy MVM extension')
> except:
>   print('Cannot load Cauchy MVM extension')
> ```

Hey @vladyorsh, thanks for the question!

I've had trouble coaxing PyTorch into working correctly when called in parallel - numerical errors happen (especially in gradients), so PyTorch is now set to run sequentially, that's (probably) why you're seeing essentially sequential performance.

This weekend  I'll try to add an experimental feature to force parallel calling (even at the cost of potential numerical errors). I'm still looking into how to solve this properly :(

**Comment by rdyro** [comment]
Hi, I've had some time to look into general multi-gpu support, but I ran out of time to take a look at your issue so far. I'm planning on taking a closer look this week.

The new multi-gpu support is here (on the main branch): [https://github.com/rdyro/torch2jax?tab=readme-ov-file#new-performant-multi-gpu-experimental](https://github.com/rdyro/torch2jax?tab=readme-ov-file#new-performant-multi-gpu-experimental) perhaps you'll find them useful.

The biggest takeaway is that I couldn't get `pmap` to work, but you should be able to obtain the same behavior (and actual performance improvements) with `shard_map`. The [data parallel](https://github.com/rdyro/torch2jax?tab=readme-ov-file#new-performant-multi-gpu-experimental) in README is quite basic but can be a good start.

I'll leave this issue open until I have a chance to look at your repro.

**Comment by rdyro** [comment]
@vladyorsh

**Comment by adam-hartshorne** [comment]
Wondering what sort of performance you got in your experiments?

I have just tried it on my very specific use case (which involves calling PyKeops inside the PyTorch func) and using 5 GPUs only gives me a 50% speed-up versus a single GPU, which is very poor speed-up. Where as, in my experience, JAX is extremely effective when sharding data across multiple GPUs using pure JAX models. 

I fear it might well be something to do with PyKeops, but thought it was worth checking.

**Comment by rdyro** [comment]
It'd be ideal to take a look at the tensorboard profile of the computation, in my simple test case (data parallel) model PyTorch correctly releases the GIL and so is able to run in parallel: [https://github.com/rdyro/torch2jax/?tab=readme-ov-file#new-performant-multi-gpu-experimental](https://github.com/rdyro/torch2jax/?tab=readme-ov-file#new-performant-multi-gpu-experimental)

Can you try running a profile of your computations? [https://jax.readthedocs.io/en/latest/profiling.html](https://jax.readthedocs.io/en/latest/profiling.html) the trace should how individual GPU utilization vs time

**Comment by adam-hartshorne** [comment]
I am not quite sure what I am looking for. The FFI appeared to being called on all the GPUS at the same time.

![Image](https://github.com/user-attachments/assets/fc8800d9-7402-4578-b3e4-235e244570bb)

**Comment by rdyro** [comment]
Cool, that's a good signal. Is it possible the library has some locking internally? The 3 rows above `jit(make_step_ode)` should be actual GPU calls (CUDA calls either from Nvidia or the library), if they never happen at the same time on multiple device, it's possible the library/PyTorch locks the calls to be sequentially internally (GIL or some form of a thread lock)

I can take a look myself if you can send the profile directory: `your-profile/plugins/profile/*`
